### DISCUSSION TITLE
Serving a pkpass directly from backend

### DISCUSSION THREAD

**Discussion by Saim-Khan1** [alexandercerutti/passkit-generator#150] [Q&A] [answered]
Hey @alexandercerutti,

It's Saim again, hope you're doing well! In your comment on [scriptpapi's YouTube video](https://www.youtube.com/watch?v=rJZdPoXHtzI&t=876s), I saw you mentioned that passes do not need to be saved to cloud storage or anything and can just be served as is. Would you be able to walk me through roughly how that works? I tried to look through the examples in this repo but I couldn't quite make sense of where I should look, I am an iOS developer primarily so would love a bit of help with this side of things :)

Assuming I have the exact same backend featured in that video, if you could please point me in the right direction I would greatly appreciate it!

Best wishes,
Saim

**Comment by alexandercerutti** [comment]
Hey there, @Saim-Khan1!
Sadly I cannot give you complete guidance because it might differ depending on many variables in play.

You can look at it this way: you are performing an HTTP Request and replying back with one or more chunks of data. Some of those chunks might be the file itself you are serving from the server. Response must have a `Content-Type`, a name, and whatever. Along with text data that describe the HTTP Response, you have some binary information. That binary information is the pass you are serving back.

You could grab that binary information, and set it to have a specific `Content-Type`, hence use it like such information type.

I could link you to an example swift app I did like 4 years ago, which takes an URL in and receives and adds the PKPass without the need of saving it first somewhere in the cloud.

Of course, the app is not updated. There might be better ways to achieve things, but it might be good enough to let you see how it might go. Do not expect high-quality / enterprise-level code. I did only a bunch of months with Swift like 4 years ago.

Here's the link: https://github.com/alexandercerutti/passkit-sample-client
And this is the code I think it might be useful to you: https://github.com/alexandercerutti/passkit-sample-client/blob/3c24467c01a2273d4c66d9a955b71141d1266161/Passkit-sample-client/ViewController.swift#L244-L272

Let me know!

**Reply by Saim-Khan1** [reply]
Sounds good, thanks a lot! I'll take a look. In terms of the code for the backend, I've temporarily made my backend repo public: https://github.com/syncstormco/PassKit-Backend
At the end of `index.js` is where I am currently uploading the .pkpass to firebase cloud storage, I was just wondering how this should be changed to send the pkpass itself as the response? If you wouldn't mind sharing the code snippet for this that would be amazing, thanks a lot :)

**Reply by alexandercerutti** [reply, marked as answer]
I don't have a code for that, but I guess you can use `response` parameter to set the `Content-Type` header (note that `passkit-generator` offers a `.mimeType` property to be used, so you won't use the wrong one), set the status code and use `.send()` and pass to it the `getAsBuffer` result.

Somehow like you can see it here: https://firebase.google.com/docs/functions/http-events?hl=en&gen=2nd#terminate_http_functions

**Reply by Saim-Khan1** [reply]
Got it, I'll give that a go and let you know. Thanks a lot for your help!

**Reply by Saim-Khan1** [reply]
Yep that worked perfectly! Thanks a lot for your help :)

**Reply by alexandercerutti** [reply]
Great to know! Thanks for your feedback! ;)

**Reply by alexandercerutti** [reply]
Now I've also fixed Serverless examples to directly return the PKPass instead of uploading on S3.

May I ask you if would you like to contribute to examples by creating a Firebase example? :)

**Reply by Saim-Khan1** [reply]
Sounds good! And sure I'd be happy to, I'm going away for a week but would be more than happy to contribute a firebase example based on my current setup when I get back :)

**Reply by alexandercerutti** [reply]
Yeah sure, feel free! I won't insist if you'll forget 😄

**Reply by Saim-Khan1** [reply]
Hey! Just put up a quick Firebase example (#153), not particularly polished and I haven't done any testing or anything with it but it's just what I have at the moment. Of course feel free to add or remove anything as you see fit :)

**Reply by alexandercerutti** [reply]
Hey @Saim-Khan1! Thank you very very much! I'm going to review it asap!

**Reply by Saim-Khan1** [reply]
Hey @alexandercerutti, sorry for the delay, I've just replied to some of your comments on the PR. I think it's going to take some further investigation!
On another note, I was wondering if you'd worked with encryption at all? Since often passes contain sensitive user information I think it would be a good idea to encrypt both requests to and responses from the server – firstly do you think it's justified or necessary? Secondly I was wondering if you had any ideas on what the best encryption method to use might be, I know Swift has the CryptoKit API for the frontend but thought I'd take your lead first in case you had worked with encryption before, especially on the backend side. Would love to hear your thoughts! (PS I can move this to another thread if you'd prefer, whatever's easiest for you)

**Reply by alexandercerutti** [reply]
Let's discuss this on the PR thread.
I've never used serious encryption here and I think this is out of the example purpose.

In my opinion, this is not really needed, even in real-world cases, when generating passes, but it actually depends on the specific case itself.
